proxy/ptunnel/commands/agent.go

```diff
@@ -0,0 +1,65 @@
+package commands
+
+import (
+	"fmt"
+	"github.com/enfabrica/enkit/lib/client"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"github.com/spf13/cobra"
+	"os"
+	"strconv"
+)
+
+type agentConfig struct {
+	Print          bool
+	SetEnv         bool
+	ListIdentities bool
+}
+
+func NewAgentCommand(bf *client.BaseFlags) *cobra.Command {
+	agentConfig := &agentConfig{}
+	c := &cobra.Command{
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return RunAgentCommand(bf, agentConfig)
+		},
+		Use:   "agent",
+		Short: "commands for the enkit specific ssh-agent",
+	}
+	// Note the following is intended to be user friendly, identities here are cert principals
+	c.Flags().BoolVarP(&agentConfig.ListIdentities, "list-identities", "l", false, "list the identities loaded current in the agent")
+	c.Flags().BoolVarP(&agentConfig.SetEnv, "set-env", "e", false, "bind the ssh-agent to the current terminal session, this allows you to interact with the enkit agent directly through ssh-add")
+	c.Flags().BoolVarP(&agentConfig.Print, "print", "p", false, "print the socket and PID of the running agent")
+	return c
+}
+
+func RunAgentCommand(bf *client.BaseFlags, config *agentConfig) error {
+	agent, err := kcerts.FindSSHAgent(bf.Local, bf.Log)
+	if err != nil {
+		return err
+	}
+	if config.Print {
+		fmt.Printf("The enkit agent is running at socket %s \n", agent.Socket)
+		fmt.Printf("The enkit agent's pid is %d \n", agent.PID)
+
+	}
+	if config.SetEnv && agent.Valid() {
+		if err := os.Setenv("SSH_AUTH_SOCK", agent.Socket); err != nil {
```

Comment:
The `SetEnv` here are a noop. In `unix` `linux`, there is no (reasonable) way for a process to change the environment variables of the calling process. Eg, `bash` started `enkit agent`, enkit agent changes its own environment, bash won't see those changes. Next command won't have the variables set.

What I would do:

- get rid of the current -e. Instead, pass `args` from the cobra `RunE` to your `RunAgentCommand`. If no option is passed (eg, no '-p', no '-l'), use `cmd/exec` to run the command specified with `args`. If `args` is empty, spawn `$SHELL`. Use `cmd.Cmd.Env` to set environment variables. Make sure to set `Cmd.Stdin` to `os.Stdin`, same for `Stdout` and `Stderr`. Try as much as possible to avoid poking at process environment variables, as those are global, affect all goroutines, and can have unpredictable effects. I've rarely seen a good reason to use Setenv in go programs. Bonus point if before running the command you print on `/dev/tty` (and it has to be `/dev/tty`) a message like 'Yoda: "Your wisdom must use, enfabrica credentials now using you are"'.
(side note: on posix systems really we should use `syscall.Exec` here, but we can leave that for later.
 
- rename `-p` to `-e`, and change the print statements above to print exactly as `ssh-agent -s` prints the output.
  (have print output the same format as ssh-agent, not a human readable string).

In terms of pseudo code, I'd reorder the operations:
- check for `-l` first. If provided, print list, return.
- check for `-e` next, if provided, print agent details, return.
- no option provided ... either run command, or shell.

Reply:
Ah see I thought it was bazel forking that was changing it, will remove

Reply:
If people want to interact with it directly, I'm all for them just copy pasting from the print statement. And we already have a way in code with the GetEnv() function to add it to our in code commands.

Reply:
Totally OK doing this in a separate PR, but was personally very excited about supporting execing a command with the env variables all set. It would make debugging and running things like `rsync` much simpler, without overriding the configuration of the current shell.